server.js

```diff
@@ -1,33 +1,160 @@
-import express from 'express'
-import bodyParser from 'body-parser'
-import cors from 'cors'
-
-// If you're using one of our datasets, uncomment the appropriate import below
-// to get started!
-// 
-// import goldenGlobesData from './data/golden-globes.json'
-// import avocadoSalesData from './data/avocado-sales.json'
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
-
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
-
-// Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
-
-// Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
-})
-
-// Start the server
+import express from "express";
+import bodyParser from "body-parser";
+import cors from "cors";
+
+import booksData from "./data/books.json";
+
+const port = process.env.PORT || 8080;
+const app = express();
+
+app.use(cors());
+app.use(bodyParser.json());
+
+app.get("/", (req, res) => {
+  res.send("Welcome to the Book Reviews [statistics] API");
+});
+
+// Authors endpoints
+// 1. List of authors & query string for author name
+app.get("/authors", (req, res) => {
+  const authors = booksData.map((item) => item.authors);
+  const uniqueAuthors = [...new Set(authors)];
+
+  // Query string - for example /authors?author=tolstoy
+  const searchedAuthor = req.query.author;
+  if (searchedAuthor) {
+    const searchedAuthorObject = booksData.filter((item) =>
+      item.authors.includes(searchedAuthor)
+    );
+    const searchedAuthorList = searchedAuthorObject.map((item) => item.authors);
+    const uniqueSearchedAuthor = [...new Set(searchedAuthorList)];
+
+    // Error when no unique searched authors found
+    if (uniqueSearchedAuthor.length === 0) {
+      const error = new Error(`${searchedAuthor} not found`);
+      error.status = 404;
+      throw error;
+    }
+    res.json(uniqueSearchedAuthor);
+  } else {
+    res.json(uniqueAuthors);
+  }
+});
+
+// 2. One specific author
+app.get("/authors/:author", (req, res, next) => {
+  const author = req.params.author;
+  const booksByAuthor = booksData.filter((item) => item.authors === author);
+  const authors = booksByAuthor.map((item) => item.authors);
+  const uniqueAuthors = [...new Set(authors)];
+
+  // Error when no unique author found
+  if (uniqueAuthors.length === 0) {
+    const error = new Error(`${author} not found`);
+    error.status = 404;
+    throw error;
+  }
+  res.json(uniqueAuthors);
+});
```

Comment:
I now, but they're not the same - the query param works more like a search-field - you can query and find any authors with the name "rowling" in them. The endpoint /:author will only give you the exact author you want - such as "J.K. Rowling".

Reply:
Good point! Maybe this should be a query param under the /books endpoint then since endpoints should be named after what they return
`/books?author=rowling`